### PR TITLE
Feature/backend/kotlin mappingy/andrew

### PR DESCRIPTION
**Pull request by AWallsRR**

### CODE REVIEW

File: Backend/mapping-kotlin/src/main/kotlin/com/staffinghub/coding/challenges/mapping/mappers/ArticleBlockDtoFactory.kt [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+class ArticleBlockDtoFactory {
+    @ObjectFactory
+    fun createArticleBlockDto(articleBlock: ArticleBlock): ArticleBlockDto {
+        return when (articleBlock) {
+            is TextBlock -> TextBlockDto(
+                text = articleBlock.text,
+                sortIndex = articleBlock.sortIndex
+            )
+            is ImageBlock -> {
+                val imageDto = articleBlock.image?.let { mapToImageDto(it) }?: ImageDto(-1, "", ImageSize.SMALL)
```
Comment (wscheffer) [inline Backend/mapping-kotlin/src/main/kotlin/com/staffinghub/coding/challenges/mapping/mappers/ArticleBlockDtoFactory.kt]: Why are we creating an additional variable only to return it, this waists memory and processing?

File: Backend/mapping-kotlin/src/main/kotlin/com/staffinghub/coding/challenges/mapping/mappers/ArticleBlockDtoFactory.kt [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+                    sortIndex = articleBlock.sortIndex
+                )
+            }
+            is VideoBlock -> VideoBlockDto(
+                url = articleBlock.url,
+                type = articleBlock.type,
+                sortIndex = articleBlock.sortIndex
+            )
+            is GalleryBlock -> GalleryBlockDto(
+                images = articleBlock.images?.mapNotNull { it?.let { image -> mapToImageDto(image) } } ?: emptyList(),
```
Comment (wscheffer) [inline Backend/mapping-kotlin/src/main/kotlin/com/staffinghub/coding/challenges/mapping/mappers/ArticleBlockDtoFactory.kt]: What is the use of the `?.let` the `mapNotNull` specifically only returns elements that aren't null?

Is you linter not telling you that the `?.let` isn't needed

File: Backend/mapping-kotlin/src/main/kotlin/com/staffinghub/coding/challenges/mapping/models/dto/blocks/GalleryBlockDto.kt [hunk 3]
```diff
@@ -5,4 +5,9 @@ import com.staffinghub.coding.challenges.mapping.models.dto.ImageDto
 data class GalleryBlockDto(
     var images: List<ImageDto>,
     override val sortIndex: Int,
-) : ArticleBlockDto
+) : ArticleBlockDto,Comparable<ArticleBlockDto>
+{
+    override fun compareTo(other: ArticleBlockDto):Int{
+        return compareValuesBy(this,other) { it.sortIndex }
```
Comment (wscheffer) [inline Backend/mapping-kotlin/src/main/kotlin/com/staffinghub/coding/challenges/mapping/models/dto/blocks/GalleryBlockDto.kt]: Please remember to format files before submitting Pull Requests.

File: Backend/mapping-kotlin/src/main/kotlin/com/staffinghub/coding/challenges/mapping/services/ArticleService.kt [hunk 4]
```diff
@@ -1,24 +1,35 @@
...
     fun articleForId(id: Long): ArticleDto {
         val article = ArticleRepository.findBy(id)
-        //TODO
-        return ArticleDto(0, "", "", "", emptyList())
+
+        if(article == null)
+            throw  ResponseStatusException( HttpStatus.NOT_FOUND, "No Article with suplied ID found")
+
+        val articleDto = mapper.mapStructMap(article)
+        return articleDto
```
Comment (wscheffer) [inline Backend/mapping-kotlin/src/main/kotlin/com/staffinghub/coding/challenges/mapping/services/ArticleService.kt]: Why create an unnecessary variable?

File: Backend/mapping-kotlin/src/test/kotlin/com/staffinghub/coding/challenges/mapping/ApplicationTests.kt [hunk 5]
```diff
@@ -1,14 +1,209 @@
...
+import com.staffinghub.coding.challenges.mapping.models.dto.blocks.VideoBlock as VideoBlockDTO
 
-@ExtendWith(SpringExtension::class)
+@ExtendWith(SpringExtension::class, MockitoExtension::class)
 @SpringBootTest
 class ApplicationTests {
+
+    private lateinit var articleController: ArticleController
+
+    private lateinit var articleService: ArticleService
```
Comment (wscheffer) [inline Backend/mapping-kotlin/src/test/kotlin/com/staffinghub/coding/challenges/mapping/ApplicationTests.kt:40]: Why aren't these being done with AutoWire?

File: Backend/mapping-kotlin/src/main/kotlin/com/staffinghub/coding/challenges/mapping/mappers/ArticleBlockDtoFactory.kt (hunk 2 above)
Comment (AWallsRR) [reply Backend/mapping-kotlin/src/main/kotlin/com/staffinghub/coding/challenges/mapping/mappers/ArticleBlockDtoFactory.kt]: image itself is a nullable item so ?.let was added to allow for the instances of the null value as well as to allow for the mapToTimageDto to run successfully.
The List Field in the GalleryBlock does not allow for any nullable Items as such if I remove the ?.let then a type mismatch error is thrown before building.

File: Backend/mapping-kotlin/src/main/kotlin/com/staffinghub/coding/challenges/mapping/mappers/ArticleBlockDtoFactory.kt (hunk 1 above)
Comment (AWallsRR) [reply Backend/mapping-kotlin/src/main/kotlin/com/staffinghub/coding/challenges/mapping/mappers/ArticleBlockDtoFactory.kt]: Apologies it has been altered to better fit best practice

File: Backend/mapping-kotlin/src/main/kotlin/com/staffinghub/coding/challenges/mapping/models/dto/blocks/GalleryBlockDto.kt (hunk 3 above)
Comment (AWallsRR) [reply Backend/mapping-kotlin/src/main/kotlin/com/staffinghub/coding/challenges/mapping/models/dto/blocks/GalleryBlockDto.kt]: Apologies I have reformatted all files

File: Backend/mapping-kotlin/src/main/kotlin/com/staffinghub/coding/challenges/mapping/services/ArticleService.kt (hunk 4 above)
Comment (AWallsRR) [reply Backend/mapping-kotlin/src/main/kotlin/com/staffinghub/coding/challenges/mapping/services/ArticleService.kt]: Apologies this was an oversite it has been altered to better fit best practice

File: Backend/mapping-kotlin/src/main/kotlin/com/staffinghub/coding/challenges/mapping/mappers/ArticleBlockDtoFactory.kt (hunk 2 above)
Comment (wscheffer) [reply Backend/mapping-kotlin/src/main/kotlin/com/staffinghub/coding/challenges/mapping/mappers/ArticleBlockDtoFactory.kt]: `images = articleBlock.images?.mapNotNull { mapToImageDto(it) } ?: emptyList(),`

Talking about the internal `it?.let`
`it` inside `mapNotNull` can't be null 

https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/map-not-null.html

image

Comment (AWallsRR) [reply Backend/mapping-kotlin/src/main/kotlin/com/staffinghub/coding/challenges/mapping/mappers/ArticleBlockDtoFactory.kt]: I understand I have made alterations for the line in questions as well as for the mapToImageDto function to allow the alterations. The issue I was having was due to the mapToImageDto not accepting Image? hence the previous iteration.

**Pull request closed without merging**